### ISSUE TITLE
[Question]how to generate SGX dcap remote attestation report for occlum image

### ISSUE THREAD

**Issue by fearlessfe** [occlum/occlum#1369]
In the development demos, an occlum instance can be build to a docker image, when I run the image. How can I get the dcap report for this occlum instance.

I want to get the dcap report for every occlum instance, so others can verify the instance

**Comment by qzheng527** [comment]
@fearlessfe Generally, mrsigner and mrenclave are used for remote verification. For these two, you could run the commands below to accquire after you finish the building of the occlum_instance.
```
cd occlum_instance
occlum print mrsigner
occlum print mrenclave
```

**Comment by fearlessfe** [comment]
@qzheng527 In the Intel's remote attestation model, an enclave can generate a REPORT;,then the REPORT can be signed and generate a QUOTE.   In the demos/remote_attestation/dcap, the example only give a simple string as a REPORT, but I cannot find a demo to get a REPORT in occlum.

In my screen scene, after a Java application built to a occlum instance, I want to get the REPORT and sign it to a QUOTE.

**Comment by qzheng527** [comment]
@fearlessfe 
https://github.com/occlum/occlum/blob/master/demos/remote_attestation/dcap/c_app/dcap_c_test.c#L76
You can use `p_quote_buffer` and `quote_size` to save the quote to a file.

**Comment by fearlessfe** [comment]
@qzheng527 but I am confused about the `report_data`. in the example, it just a simple string, what the `report_data` should be?  Do you mean the `mrsigner` and `mrenclave` has been in the `p_quote_buffer` and the `report_data` can be anything?

```
    sgx_report_data_t report_data = { 0 };
    char *data = "ioctl DCAP report data example";
    memcpy(report_data.d, data, strlen(data));

    // Get the Quote
    ret = dcap_generate_quote(handle, p_quote_buffer, &report_data);
```

**Comment by qzheng527** [comment]
> @qzheng527 but I am confused about the `report_data`. in the example, it just a simple string, what the `report_data` should be? Do you mean the `mrsigner` and `mrenclave` has been in the `p_quote_buffer` and the `report_data` can be anything?
> 
> ```
>     sgx_report_data_t report_data = { 0 };
>     char *data = "ioctl DCAP report data example";
>     memcpy(report_data.d, data, strlen(data));
> 
>     // Get the Quote
>     ret = dcap_generate_quote(handle, p_quote_buffer, &report_data);
> ```

`report_data` has the structure of `sgx_report_data_t` which can be seen at 
https://github.com/occlum/incubator-teaclave-sgx-sdk/blob/13fb3476f53d21614bb863dc7de825e70938ce36/sgx_types/src/types.rs#L424

It is a data array (64 bytes). You can put anything in it, including string or hex, as long as the size is not overflow.

**Comment by LearningofPython** [comment]
> @fearlessfe All the flow you mentioned could be done by minor changes on the [demo](https://github.com/occlum/occlum/blob/master/demos/remote_attestation/dcap). https://github.com/occlum/occlum/blob/master/demos/remote_attestation/dcap/c_app/dcap_c_test.c#L76 You can use `p_quote_buffer` and `quote_size` to save the quote to a file.

// Save the Quote to a file
    FILE *file = fopen("quote.bin", "wb");
    if (file == NULL) {
        printf("Couldn't open file to save quote\n");
        free(p_quote_buffer);
        return;
    }
    fwrite(p_quote_buffer, 1, quote_size, file);
    fclose(file);
    printf("Quote saved to file successfully\n");
    This is my code.I have saved the quote to a file,and see "Quote saved to file successfully",but I can't see the quote file.Why？Where is it?

**Comment by qzheng527** [comment]
@fearlessfe By detault all the FS operation in Occlum is in the SEFS which  couldn't be seen outside of TEE.  You can add code to read the quote file, or just write it in the directory "/host/" which it is mounted as host FS which could be seen in host side in the directory of `occlum_instance`.

**Comment by LearningofPython** [comment]
> @fearlessfe By detault all the FS operation in Occlum is in the SEFS which couldn't be seen outside of TEE. You can add code to read the quote file, or just write it in the directory "/host/" which it is mounted as host FS which could be seen in host side in the directory of `occlum_instance`.

Thank you.I have another question.I want to place the key in the image directory, then run "occlum build", and then have the Java code read this key. Is this approach secure, and can it achieve the goal of storing the key in the Enclave?

**Comment by qzheng527** [comment]
> > @fearlessfe By detault all the FS operation in Occlum is in the SEFS which couldn't be seen outside of TEE. You can add code to read the quote file, or just write it in the directory "/host/" which it is mounted as host FS which could be seen in host side in the directory of `occlum_instance`.
> 
> Thank you.I have another question.I want to place the key in the image directory, then run "occlum build", and then have the Java code read this key. Is this approach secure, and can it achieve the goal of storing the key in the Enclave?

No. The occlum build by default doesn't generate encrypted image. Thus putting key in the image directory before "occlum build" can't protect the key. Users with deep understanding of Occlum FS could dump the key out. But if you generate the key and save to FS when running some code in Occlum, it is safe and sealed.
To generated encrypted Occlum image, you can do occlum build with option "--image-key" and get the key through RA when start the Occlum. 
https://occlum.readthedocs.io/en/latest/remote_attestation.html#init-ra-solution

**Comment by LearningofPython** [comment]
> > > @fearlessfe By detault all the FS operation in Occlum is in the SEFS which couldn't be seen outside of TEE. You can add code to read the quote file, or just write it in the directory "/host/" which it is mounted as host FS which could be seen in host side in the directory of `occlum_instance`.
> > 
> > 
> > Thank you.I have another question.I want to place the key in the image directory, then run "occlum build", and then have the Java code read this key. Is this approach secure, and can it achieve the goal of storing the key in the Enclave?
> 
> No. The occlum build by default doesn't generate encrypted image. Thus putting key in the image directory before "occlum build" can't protect the key. Users with deep understanding of Occlum FS could dump the key out. But if you generate the key and save to FS when running some code in Occlum, it is safe and sealed. To generated encrypted Occlum image, you can do occlum build with option "--image-key" and get the key through RA when start the Occlum. https://occlum.readthedocs.io/en/latest/remote_attestation.html#init-ra-solution

OK
